src/TwigFactory.php

```diff
@@ -44,4 +47,42 @@ public static function newFromConfig( array $config, WikiContentProvider $provid
 
 		return $twig;
 	}
+
+	public function newWikiPageLoader( WikiContentProvider $provider ) {
+		if ( empty( $this->config['loaders']['wiki'] ) ) {
+			return null;
+		}
+		return new TwigPageLoader( $provider );
+	}
+
+	public function newFileSystemLoader() {
+		if ( empty( $this->config['loaders']['filesystem'] ) ) {
+			return null;
+		}
+		if ( empty( $this->config['loaders']['filesystem']['template-dir'] ) ) {
+			$templateDir = [ 'app/templates' ];
+		}
+		elseif( is_string( $this->config['loaders']['filesystem']['template-dir'] ) ) {
+			$templateDir = [ $this->config['loaders']['filesystem']['template-dir'] ];
+		}
+		elseif( is_array( $this->config['loaders']['filesystem']['template-dir'] ) ) {
+			$templateDir = $this->config['loaders']['filesystem']['template-dir'];
+		}
+		else {
+			throw new \RuntimeException( 'wrong template dir type' );
+		}
+		$appRoot = realpath( __DIR__ . '/..' ) . '/';
+		$templateDir = array_map( function( $dir ) use ( $appRoot ) {
+			if ( strlen( $dir ) == 0 || $dir{0} != '/' ) {
+				return $appRoot . $dir;
+			}
+			return $dir;
+		}, $templateDir );
+		return new Twig_Loader_Filesystem( $templateDir );
+	}
+
+	public function newArrayLoader() {
+		$templates = $this->config['loaders']['array'] ?? [];
```

Comment:
Right. Btw, do we have a list somewhere of PHP7 features we still have to use in this project? Gotta use 'em all!